### ISSUE TITLE
fusile hanging on read

### ISSUE THREAD

**Issue by unwitting** [Munter/fusile#19]
My simple use-case:

```
$ ll src/
total 16
drwxr-xr-x  4 jackpreston  staff  136 25 Sep 16:20 .
drwxr-xr-x  5 jackpreston  staff  170 25 Sep 16:20 ..
-rw-r--r--  1 jackpreston  staff   32 25 Sep 16:16 test.es6.js
-rw-r--r--  1 jackpreston  staff   53 25 Sep 16:19 test.scss

$ ./node_modules/.bin/fusile src/ srv
Transpilers loaded:  babel, node-sass
File system started at /Users/jackpreston/tmp/srv
To stop it, press Ctrl+C

(another terminal)
$ cat srv/test.css

-- fusile hangs here for ~1min --

cat: srv/test.css: Socket is not connected
```

Told you I'd let you know :D any ideas?

**Comment by unwitting** [comment]
Verbose output shows we _do_ notice the read, but that's where it stops, so I'm assuming this is specific to the sass flow.

```
tmp $ ./node_modules/.bin/fusile src/ srv -v
Transpilers loaded:  babel, node-sass
File system started at /Users/jackpreston/tmp/srv
To stop it, press Ctrl+C
getattr /
getxattr /
getxattr /
getxattr /
getattr /.hidden
getxattr /
getxattr /
getxattr /
getattr /
getxattr /
getxattr /
getxattr /
getxattr /
getxattr /
getxattr /
getxattr /
getattr /.hidden
getxattr /
getxattr /
getxattr /
getxattr /
getxattr /
getxattr /
getattr /
getxattr /
getxattr /
getattr /
getattr /
readdir /
getattr /
readdir /
getattr /
getattr /test.css
getattr /test.css
getattr /test.css
open /test.css
read /test.css
```

**Comment by Munter** [comment]
There must be a missing `.catch()` somewhere. Darn promises. What you should see is at least a stylesheet with a very obvious error instead of just hanging

**Comment by Munter** [comment]
What's the content of `test.scss`?

**Comment by unwitting** [comment]
```
$color: red;

body {
    background-color: $color;
}
```

I think is valid?

**Comment by Munter** [comment]
Looks valid to me. I'll see if I can reproduce

**Comment by unwitting** [comment]
Right, much logging later,the culprit is `adapter.renderFile(pathName, transpilerOptions)` for the node-sass adapter not resolving in the `continuation = adapter.renderFile(pathName, transpilerOptions).then...` line of tolk.js

**Comment by Munter** [comment]
Ah, not catching accord failures. What version of node-sass do you have installed?

**Comment by unwitting** [comment]
Looks that way - I've tried sticking a `.catch()` in just after it to make sure, and it's not getting hit. I'm on node-sass 3.3.3. Potentially accord has trouble working fully with this version, I'll go ahead and try an earlier one.

**Comment by unwitting** [comment]
Works like a charm with node-sass 2.1.1 :) gonna do a bit of a binary chop

**Comment by unwitting** [comment]
node-sass 3.0.0 fails, 2.x succeeds.

**Comment by unwitting** [comment]
Turns out node-sass moved from options.success callbacks to node-style render(..., cb) style callbacks in v3 - accord doesn't make use of this until jenius/accord@1b3cf61e (0.20.1) - testing out bumping accord version from tolk.

**Comment by Munter** [comment]
Getting tired of accord not being multible major version compatible. I think it might make sense to drop accord for node-sass and just use my own compatibility wrapper: https://github.com/Munter/node-sass-evergreen

**Comment by unwitting** [comment]
That would enable me to have both learnt about and forgotten about accord in one day.

**Comment by Munter** [comment]
@jenius Accord is still hard coding one specific major version compatible engine instead of actually using the concept that is supports multiple. I think https://github.com/jenius/accord/blob/master/lib/adapters/scss/index.coffee should look more like https://github.com/Munter/node-sass-evergreen/blob/master/lib/index.js

Anything else will always get consumers into trouble in the intersection of time where multiple major versions are in use. It shouldn't be accords (or tools based on it) responsibility to force developers to upgrade their transpiler version

**Comment by Munter** [comment]
So it looks like two fixes are needed.
- Catch transpiler adapter errors from accord in tolk and do something reasonable with them so fuse doesn't lock up.
- Be better at supporting real world versions of transpiler engines without throwing

**Comment by unwitting** [comment]
Problem here is that accord isn't throwing anything *to* catch - simply no return ever happens since it wasn't giving sass the right cb format.

**Comment by Munter** [comment]
So maybe I need a catch all timeout that just aborts whatever was running if it took too long

**Comment by unwitting** [comment]
Yeah that would work - I'd not be that happy with it if it wasn't for the fact that you're obviously thinking of moving to a better (?) way of managing the versions of compiler adapters in the near future, since it brings all the classic issues of "how long is too long?", "what if I have an absolutely massive and hard to compile sass file?" etc. But as an interim solution it would at least stop things hanging mysteriously!

**Comment by unwitting** [comment]
Side note, now I've got it working I love this tool as much as I originally thought I did, very good idea :)

**Comment by jenius** [comment]
Hey there!

Sorry, I'm not entirely sure I understand what the issue is here. Are you getting an accord failure, but it's not throwing? Could you drop me a code sample I can use to reproduce this? Would be happy to patch it up.

As far as version compatibility, it's all there. It works with `2.x` and `3.x`. It seems like the accord version was out of date and you were using a version from before the release of `node-sass@3.x`. I assume the fix submitted here will work perfectly fine.

Let me know if there's anything else I can help with here!

**Comment by unwitting** [comment]
No real problem here now, the issue was that accord 18 didn't support
node-sass 3 callback format and thus wasn't resolving ever. Upgrade to 20
fixes it in this instance.

**Comment by jenius** [comment]
Cool, glad this worked out! From version 20 forward, accord throws a very explicit and specific error for unsupported versions, so hopefully this will be much easier to catch from here out :grinning:

**Comment by unwitting** [comment]
Hopefully! I guess it's part of the responsibility of a universal API :D

Are supported versions whitelisted? If `node-sass` magicks up a v4 out of nowhere will accord throw an error against it until you guys have verified that it works with it?

**Comment by jenius** [comment]
Yes, it will. I figured this was the most responsible way to handle it. Even if the major version bump doesn't break anything, better to vet and run a quick release than let it ship out broken. I could be argued the other way though if someone presented a compelling argument!

**Comment by unwitting** [comment]
> it is not possible for any language to break with accord now, unless they break semver
this ^ :)

**Comment by unwitting** [comment]
@Munter any chance of an npm publish of tolk to finish off the accord version upgrade?

**Comment by unwitting** [comment]
(I don't mean to hassle, promise :D)

**Comment by Munter** [comment]
tolk 1.1.2 is out with the accord updates. But my unit tests are not passing on fusile. Looking at it now

**Comment by Munter** [comment]
I just released fusile 1.0.3

Does this solve the problem?

**Comment by unwitting** [comment]
It does indeed. I see `npm test` now works out of the box - that's awesome, I can help out much more easily with them now :) I'll close this as fixed.
